routing/record/validation.go

```diff
@@ -100,3 +112,18 @@ func CheckRecordSig(r *pb.Record, pk ci.PubKey) error {
 	}
 	return nil
 }
+
+type RecordSorter []*pb.Record
+
+func (rs RecordSorter) Less(i, j int) bool {
+	mv, _ := MoreValid(rs[j], rs[i])
```

Comment:
On Wed, Apr 01, 2015 at 03:47:41PM -0700, Juan Batiz-Benet wrote:
> The problem with using a timestamp for every record is that
> timestamps dont mean anything in some contexts. for example, if i
> use a git-chain style approach, the last record could be a year old
> and still be the intended (most valid) one. A timestamp here would
> make you chose the wrong name. we also get into clock skew problems,
> and so on...

I'm new here, so take this with a grain of salt ;).  But DHT records
are all per-node things right (more details on my possibly incorrect
understanding in [1])?  The per-node namespacing in the DHT keyspace
is how you avoid name collisions.  So if you want a validity test, you
just have to sign all the key/payload associations:

  DHT key: /{per-node-thing-you-want}/{node-you're-asking-about}
  DHT value: {
       payload: {some-data},
       signature: {signature(
         {DHT-key} + ␀ + {payload-serialized-to-protobuf},
         {key-for-node-you're-asking-about}),
    }

Then you just check the signature to ensure that the payload is valid
for that key (says the node you're asking about).

You can still have multiple valid results though (if the node mutated
the value in the tree, and the new value hasn't finished propagating).
For deciding between valid values, any kind of increasing counter
(timestamp, serial counter, …) would be fine, and picking one to use
globally at the protocol level shouldn't restrict users in any way
that they'd care about.  Personally, I think timestamps are a better
choice, since then:

* You get an idea of how recent the record is (although as you say, an
  old record may still be the most current).
* You don't have to keep track of the serial number of your last
  record (most systems running IPFS will already be keeping track of
  the time)

Things like clock skew won't matter much, since it's unlikely that a
given node will have such a skewy clock that it publishes a new record
with a timestamp from before it's last record.  This *could* happen
for folks who run rare clock syncs, but I think the fraction of
effected records will be teensy, and the only negative effect is that
some folks might get the stale-but-higher-timestamped record until
your next publication.

> … also timestamps on every record blow up the size of the record
> space…

No getting around that though ;).  How big a deal is it?

[1]: https://github.com/ipfs/go-ipfs/issues/1045#issuecomment-94200028